File: solutions/day2.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const report = raw.split("\n").map((level) => level.split(" ").map(Number));

const levels = report.map((level) => {
  const diffs = [];
  for (let i = 0; i < level.length - 1; i++) {
    diffs.push(level[i] - level[i + 1]);
  }
  return diffs;
});

const { countPart1, countPart2 } = levels.reduce<{
  countPart1: number;
  countPart2: number;
}>(
  (acc, diffs) => {
    const { countPos, countNeg, countBadDiff } = diffs.reduce<{
      countPos: number;
      countNeg: number;
      countBadDiff: number;
    }>(
      (acc, diff) => {
        if (diff > 0) {
          acc.countPos++;
        } else {
          acc.countNeg++;
        }
        if (Math.abs(diff) > 3 || Math.abs(diff) === 0) {
          acc.countBadDiff++;
        }
        return acc;
      },
      { countPos: 0, countNeg: 0, countBadDiff: 0 }
    );

    // Part 1
    if (countBadDiff === 0 && (countPos === 0 || countNeg === 0)) {
      acc.countPart1++;
    }

    // Part 2
    const conditions = [
      countBadDiff <= 1 && (countPos === 0 || countNeg === 0),
      countBadDiff === 0 &&
        ((countPos === 1 && countNeg > 0) || (countNeg === 1 && countPos > 0)),
    ];
    if (conditions.some(Boolean)) {
      acc.countPart2++;
    }
    return acc;
  },
  { countPart1: 0, countPart2: 0 }
);

console.log("Part 1:", countPart1);
console.log("Part 2:", countPart2);

File: solutions/python/day2.py
with open("./input.txt", "r") as txt:
  rep: str = txt.read().rstrip()
  txt.close()

report = [list(map(int, _.split(' '))) for _ in rep.split('\n')]

levels: list[list[int]] = []
for i in range(len(report)):
  diffs = [0] * (len(report[i]) - 1)
  for j in range(len(report[i]) - 1):
    diffs[j] = report[i][j] - report[i][j + 1]
  levels.append(diffs)

countPart1, countPart2 = 0, 0
for diffs in levels:
  countPos, countNeg, countBadDiff = 0, 0, 0
  for diff in diffs:
    if diff > 0:
      countPos += 1
    else:
      countNeg += 1
    if abs(diff) > 3 or abs(diff) == 0:
      countBadDiff += 1
  
  if countBadDiff == 0 and (countPos == 0 or countNeg == 0):
    countPart1 += 1

  conditions = [
    countBadDiff == 0 and (countPos == 0 or countNeg == 0),
    countBadDiff == 1 and (countPos == 0 or countNeg == 0),
    countBadDiff == 0 and ((countPos == 1 and countNeg > 0) or (countNeg == 1 and countPos > 0))
  ]
  if conditions.count(True) == 1:
    countPart2 += 1

print("Part 1:", countPart1)
print("Part 2:", countPart2)
